boa/src/object/gcobject.rs

```diff
@@ -176,31 +206,109 @@ impl GcObject {
                             arguments_obj,
                             context,
                         )?;
+                        // push the environment first so that it will be used by default parameters
+                        context.push_environment(local_env.clone());
+
+                        // Add argument bindings to the function environment
+                        for (i, param) in params.iter().enumerate() {
+                            // Rest Parameters
+                            if param.is_rest_param() {
+                                function.add_rest_param(param, i, args, context, &local_env);
+                                break;
+                            }
 
-                        context.push_environment(local_env);
+                            has_parameter_expressions =
+                                has_parameter_expressions || param.init().is_some();
+
+                            let value = match args.get(i).cloned() {
+                                None | Some(Value::Undefined) => {
+                                    if let Some(init) = param.init() {
+                                        init.run(context).unwrap_or(Value::Undefined)
+                                    } else {
+                                        Value::Undefined
+                                    }
```

Comment:
I think this can be simplified with a `map()` and `unwrap_or_default()`, something along the lines of:
`parma.init().map(|init| init.run(context)).flatten().unwrap_or_default()`